### PR TITLE
Implemented query search for client

### LINKED ISSUE #59: LocalBackend: Implement Search
### What is the expected behavior?

Implement the search in the localbackend.

### PR DESCRIPTION
**Pull request by AnsahMohammad**
### Summary

Implemented query search for client

- [x] Implemented Search
- [x] Added test in test_backend

### Details and comments
Currently, uses `trial.name.find(query)` to filter out all trials that contain the substring 'query' in their names.

### CODE REVIEW

File: client/purplecaffeine/core.py [hunk 1]
```diff
@@ -471,6 +471,9 @@ def list(
             with open(path, "r", encoding="utf-8") as trial_file:
                 trial_dict = json.load(trial_file, cls=TrialDecoder)
                 trials.append(Trial(**trial_dict))
+
+        if query:
```
Comment (mickahell) [inline client/purplecaffeine/core.py:475]: The problem here is you applied your query after the limit offset. We should be able to apply the limit to the list return by the query

File: client/tests/test_backend.py [hunk 2]
```diff
@@ -35,9 +35,12 @@ def test_save_get_list_local_storage(self):
         with self.assertRaises(ValueError):
             self.local_storage.get(trial_id="999")
         # List
-        list_trials = self.local_storage.list()
+        list_trials = self.local_storage.list(query="keep_trial")
         self.assertTrue(isinstance(list_trials, list))
         self.assertTrue(isinstance(list_trials[0], Trial))
+        list_trials = self.local_storage.list(query="trial999")
```
Comment (mickahell) [inline client/tests/test_backend.py]: We should test that we get a response that nothing was found by a message or raise an error. Up to you to choose :)

**Review by mickahell — CHANGES_REQUESTED** [review]
Some tricks/logics to apply but the `find` seems to be a good way to go :)

I'm wondering maybe, having query apply on the field `tag` could also be usefull

**Review by mickahell — COMMENTED** [review]
Good to me :) just don't we wanted to also have the query applied to the tags ?

**Review by mickahell — APPROVED** [review]

**Review by IceKhan13 — APPROVED** [review]
another hit for UX 🚀

**Comment by IceKhan13** [comment]
@AnsahMohammad can you resolve conflict?

**Comment by AnsahMohammad** [comment]
@IceKhan13 extended the query to include search on tags and description.

File: client/purplecaffeine/core.py [hunk 3]
```diff
@@ -465,12 +465,26 @@ def list(
...
+            trials = [
+                trial
+                for trial in trials
+                if (query in trial.tags)
+                or (trial.name.find(query) != -1)
+                or (trial.description.find(query) != -1)
+            ]
+            if len(trials) == 0:
+                logging.warning("No Trials returned for the query '%s'.", query)
```
Comment (IceKhan13) [inline client/purplecaffeine/core.py]: I think we should return empty results if nothing has been found :) 

Error will scare people :)

**Review by IceKhan13 — CHANGES_REQUESTED** [review]

**Comment by AnsahMohammad** [comment]
Okay, fixed that 😄

**Comment by IceKhan13** [comment]
tests are failing :)

**Comment by AnsahMohammad** [comment]
@IceKhan13 Fixed tests! There was a typo!

**Review by IceKhan13 — APPROVED** [review]
Merging!

**Pull request merged**